consensus/src/aggregation/engine.rs

```diff
@@ -0,0 +1,781 @@
+                    self.pending.iter_mut().for_each(|(_, pending)| {
+                        match pending {
+                            Pending::Unverified(acks) => {
+                                acks.retain(|epoch, _| *epoch >= min_epoch);
+                            }
+                            Pending::Verified(_, acks) => {
+                                acks.retain(|epoch, _| *epoch >= min_epoch);
+                            }
+                        }
+                    });
+
+                    continue;
+                },
+
+                // Sign a new ack
+                request = self.digest_requests.next_completed() => {
+                    let DigestRequest { index, result, timer } = request;
+                    drop(timer); // Record metric. Explicitly reference timer to avoid lint warning
+                    match result {
+                        Err(err) => {
+                            warn!(?err, ?index, "automaton returned error");
+                            self.metrics.digest.inc(Status::Dropped);
+                        }
+                        Ok(digest) => {
+                            if let Err(err) = self.handle_digest(index, digest, &mut net_sender).await {
+                                warn!(?err, ?index, "handle_digest failed");
+                                continue;
+                            }
+                        }
+                    }
+                },
+
+                // Handle incoming acks
+                msg = net_receiver.recv() => {
+                    // Error handling
+                    let (sender, msg) = match msg {
+                        Ok(r) => r,
+                        Err(err) => {
+                            warn!(?err, "ack receiver failed");
+                            break;
+                        }
+                    };
+                    let mut guard = self.metrics.acks.guard(Status::Invalid);
+                    let peer_ack = match msg {
+                        Ok(peer_ack) => peer_ack,
+                        Err(err) => {
+                            warn!(?err, ?sender, "ack decode failed");
+                            continue;
+                        }
+                    };
+                    let PeerAck { ack, tip } = peer_ack;
+
+                    // Update the tip manager
+                    if self.safe_tip.update(sender.clone(), tip).is_some() {
+                        // Fast-forward our tip if needed
+                        let safe_tip = self.safe_tip.get();
+                        if safe_tip > self.tip {
+                           self.fast_forward_tip(safe_tip).await;
+                        }
+                    }
+
+                    // Validate that we need to process the ack
+                    if let Err(err) = self.validate_ack(&ack, &sender) {
+                        warn!(?err, ?sender, "ack validate failed");
```

Comment:
nit: block here if we get ack from wrong peer or if signature is invalid?

Reply:
https://github.com/commonwarexyz/monorepo/pull/974/commits/a665cf0993d2fbc69f079282e00fee4a5b69acb9